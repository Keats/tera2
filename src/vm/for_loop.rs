use crate::value::{Key, StringKind};
use crate::Value;

use serde::ser::SerializeStruct;
use serde::{Serialize, Serializer};
use std::collections::{BTreeMap, VecDeque};
use std::sync::Arc;

// TODO: perf improvements, less to_string

/// Enumerates on the types of values to be iterated, scalars and pairs
#[derive(Debug, PartialEq)]
pub enum ForLoopValues {
    /// Values for an array style iteration
    Array(VecDeque<Value>),
    Bytes(VecDeque<u8>),
    // TODO: use unic-segment as a feature and use graphemes rather than char
    /// Values for a per-character iteration on a string
    String(VecDeque<char>),
    /// Values for an object style iteration
    Object(VecDeque<(Key, Value)>),
}

impl ForLoopValues {
    #[inline(always)]
    pub fn pop_front(&mut self) -> (Value, Value) {
        match self {
            ForLoopValues::Array(a) => (Value::Null, a.pop_front().unwrap()),
            ForLoopValues::Bytes(a) => (Value::Null, Value::U64(a.pop_front().unwrap() as u64)),
            ForLoopValues::String(a) => (
                Value::Null,
                Value::String(
                    Arc::new(a.pop_front().unwrap().to_string()),
                    StringKind::Normal,
                ),
            ),
            ForLoopValues::Object(a) => {
                let (key, value) = a.pop_front().unwrap();
                (key.into(), value)
            }
        }
    }

    #[inline(always)]
    pub fn len(&self) -> usize {
        match self {
            ForLoopValues::Array(a) => a.len(),
            ForLoopValues::Bytes(a) => a.len(),
            ForLoopValues::String(a) => a.len(),
            ForLoopValues::Object(a) => a.len(),
        }
    }

    #[inline(always)]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

#[derive(Debug, Eq, PartialEq)]
struct Loop {
    index: usize,
    index0: usize,
    first: bool,
    last: bool,
    length: usize,
}

impl Loop {
    fn advance(&mut self) {
        self.index += 1;
        self.index0 += 1;
        self.first = false;
        self.last = self.index == self.length;
    }
}

impl Serialize for Loop {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("Loop", 5)?;
        s.serialize_field("index", &self.index)?;
        s.serialize_field("index0", &self.index0)?;
        s.serialize_field("first", &self.first)?;
        s.serialize_field("last", &self.last)?;
        s.serialize_field("length", &self.length)?;
        s.end()
    }
}

#[derive(Debug, PartialEq)]
pub(crate) struct ForLoop {
    values: ForLoopValues,
    loop_data: Loop,
    pub(crate) end_ip: usize,
    context: BTreeMap<String, Value>,
    value_name: Option<String>,
    key_name: Option<String>,
    current_values: (Value, Value),
}

impl ForLoop {
    pub fn new(is_key_value: bool, container: Value) -> Self {
        // TODO: keep an iterator instead of that thing
        let values = match container {
            Value::Map(map) => {
                if !is_key_value {
                    todo!("Error");
                }
                let mut vals = VecDeque::with_capacity(map.len());
                for (key, val) in map.as_ref() {
                    vals.push_back((key.clone(), val.clone()));
                }
                ForLoopValues::Object(vals)
            }
            Value::String(s, _) => {
                if is_key_value {
                    todo!("Error");
                }

                let chars: VecDeque<_> = s.chars().collect();
                ForLoopValues::String(chars)
            }
            Value::Bytes(b) => {
                if is_key_value {
                    todo!("Error")
                }
                let mut bytes = VecDeque::with_capacity(b.len());
                for a in b.iter() {
                    bytes.push_back(*a);
                }
                // TODO: add tests to loops on bytes
                ForLoopValues::Bytes(bytes)
            }
            Value::Array(arr) => {
                if is_key_value {
                    todo!("Error");
                }
                let mut vals = VecDeque::with_capacity(arr.len());
                for a in arr.iter() {
                    vals.push_back(a.clone());
                }
                ForLoopValues::Array(vals)
            }
            _ => todo!("handle error"),
        };

        let length = values.len();
        let loop_data = Loop {
            index: 1,
            index0: 0,
            first: true,
            last: length == 0,
            length,
        };

        Self {
            values,
            loop_data,
            end_ip: 0,
            context: BTreeMap::new(),
            value_name: None,
            key_name: None,
            current_values: (Value::Null, Value::Null),
        }
    }

    pub(crate) fn store_local(&mut self, name: &str) {
        if self.value_name.is_none() {
            self.value_name = Some(name.to_string());
        } else if self.key_name.is_none() {
            self.key_name = Some(name.to_string());
        }
    }

    /// Advance the counter only after the end ip has been set (eg we start incrementing only from the
    /// second time we see the loop)
    pub(crate) fn advance(&mut self) {
        if self.end_ip == 0 {
            self.current_values = self.values.pop_front();
            return;
        } else {
            self.loop_data.advance();
            self.context.clear();
            self.current_values = self.values.pop_front();
        }
    }

    #[inline(always)]
    pub(crate) fn is_over(&self) -> bool {
        self.values.is_empty()
            || (self.end_ip != 0 && self.loop_data.index0 == self.loop_data.length - 1)
    }

    pub(crate) fn iterated(&self) -> bool {
        self.loop_data.index0 > 0
    }

    pub(crate) fn clear_context(&mut self) {
        self.context.clear();
    }

    pub(crate) fn store(&mut self, name: &str, value: Value) {
        self.context.insert(name.to_string(), value);
    }

    pub(crate) fn get(&self, name: &str) -> Option<Value> {
        // Special casing the loop variable
        if name == "loop" {
            return Some(Value::from_serializable(&self.loop_data));
        }

        if self.value_name.as_deref() == Some(name) {
            return Some(self.current_values.1.clone());
        }

        if self.key_name.as_deref() == Some(name) {
            return Some(self.current_values.0.clone());
        }

        self.context.get(name).cloned()
    }
}
