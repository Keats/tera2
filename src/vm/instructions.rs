use crate::utils::Span;
use crate::value::Value;
use std::collections::HashMap;
use std::fmt;
use std::fmt::Formatter;

#[derive(Debug, Clone)]
pub enum Instruction {
    /// Pushing a value to the stack
    LoadConst(Value),
    /// Reading a variable/function
    LoadName(String),
    /// Write the raw string given
    WriteText(String),
    /// Writes the value on the top of the stack
    WriteTop,
    /// Set the last value on the stack in the current context
    Set(String),
    /// Set the last value on the stack in the global context. Same as Set outside of loops.
    SetGlobal(String),
    /// Include the given template
    Include(String),
    /// Get the named field of the top stack value (`person.name`)
    LoadAttr(String),
    /// Handles `a[b]`. `b` is the top stack value, `a` the one before
    BinarySubscript,

    /// Create a map for the kwargs of a function. Inner field is the number of values
    BuildKwargs(usize),
    /// Create a list. Inner field is the number of values
    BuildList(usize),
    /// Calls the named Tera function
    CallFunction(String),
    /// Calls the macro in the given namespace. (namespace_idx, name_idx)
    CallMacro(usize, usize),
    /// Apply the given filter
    ApplyFilter(String),
    RunTest(String),

    // math
    Mul,
    Div,
    FloorDiv,
    Mod,
    Plus,
    Minus,
    Power,

    // logic
    LessThan,
    GreaterThan,
    LessThanOrEqual,
    GreaterThanOrEqual,
    Equal,
    NotEqual,

    StrConcat,
    In,

    // unary
    Not,
    Negative,
}

#[derive(Clone)]
pub struct Chunk {
    instructions: Vec<Instruction>,
    /// instruction idx -> Span
    spans: HashMap<u32, Span>,
    // What should it be there? Template name?
    name: String,
}

impl Chunk {
    pub(crate) fn new(name: &str) -> Self {
        Self {
            instructions: Vec::with_capacity(256),
            spans: HashMap::with_capacity(256),
            name: name.to_owned(),
        }
    }

    pub(crate) fn add(&mut self, instr: Instruction) -> u32 {
        let idx = self.instructions.len();
        self.instructions.push(instr);
        idx as u32
    }

    pub(crate) fn add_instruction_with_span(&mut self, instr: Instruction, span: Span) {
        let idx = self.add(instr);
        self.spans.insert(idx, span);
    }
}

impl fmt::Debug for Chunk {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        writeln!(f, "=== {} ===", self.name)?;

        // let mut offset = 0;
        for (offset, instr) in self.instructions.iter().enumerate() {
            writeln!(f, "{offset:>04} {instr:?}")?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_size() {
        assert_eq!(std::mem::size_of::<Instruction>(), 32);
    }

    // #[test]
    // fn test_debug() {
    //     let mut chunk = Chunk::new("hello");
    //     chunk.add_instruction(Instruction::LoadConst(Value::U64(1)));
    //     chunk.add_instruction(Instruction::LoadVar("ho".to_string()));
    //     println!("{:?}", chunk);
    //     assert!(false);
    // }
}
