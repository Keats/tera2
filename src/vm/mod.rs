mod for_loop;
mod stack;
mod state;

pub(crate) mod interpreter;
