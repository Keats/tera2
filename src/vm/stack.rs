use crate::Value;

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) struct Stack {
    values: Vec<Value>,
}

impl Stack {
    pub(crate) fn new() -> Self {
        Self {
            // TODO: check the size of the stack of an average template
            values: Vec::with_capacity(64),
        }
    }

    pub(crate) fn push(&mut self, val: Value) {
        self.values.push(val);
    }

    pub(crate) fn pop(&mut self) -> Value {
        self.values.pop().expect("to have a value")
    }

    pub(crate) fn peek(&mut self) -> &Value {
        self.values.last().expect("to peek a value")
    }
}
