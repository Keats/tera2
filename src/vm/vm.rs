use crate::template::Template;

struct VirtualMachine<'t> {
    template: &'t Template,
    ip: usize,
}

impl<'t> VirtualMachine<'t> {
    pub fn new(template: &'t Template) -> Self {
        Self { template, ip: 0 }
    }

    fn interpret(&self) {}
}
