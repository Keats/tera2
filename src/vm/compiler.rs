//! AST -> bytecode
use crate::parsing::ast::{BinaryOperator, Expression, Node, UnaryOperator};
use crate::vm::instructions::{Chunk, Instruction};

// TODO: a bit weird to have the compiler in the vm part?
pub(crate) struct Compiler {
    pub(crate) chunk: Chunk,
    body: String,
    /// How many bytes of raw content we've seen
    bytes_size: usize,
}

impl Compiler {
    pub(crate) fn new(name: &str, body: &str) -> Self {
        Self {
            chunk: Chunk::new(name),
            body: body.to_string(),
            bytes_size: 0,
        }
    }

    fn compile_expr(&mut self, expr: Expression) {
        match expr {
            Expression::Const(e) => {
                let (val, _) = e.into_parts();
                self.chunk.add_instruction(Instruction::LoadConst(val));
            }
            Expression::Array(_) => {}
            Expression::Var(e) => {
                let (val, _) = e.into_parts();
                self.chunk.add_instruction(Instruction::LoadVar(val.name));
            }
            Expression::GetAttr(_) => {}
            Expression::GetItem(_) => {}
            Expression::Test(_) => {}
            Expression::MacroCall(_) => {}
            Expression::Filter(_) => {}
            Expression::FunctionCall(_) => {}
            Expression::UnaryOperation(e) => {
                let (op, _) = e.into_parts();
                self.compile_expr(op.expr);
                match op.op {
                    UnaryOperator::Not => self.chunk.add_instruction(Instruction::Not),
                    UnaryOperator::Minus => self.chunk.add_instruction(Instruction::Negative),
                };
            }
            Expression::BinaryOperation(e) => {
                let (op, span) = e.into_parts();
                // TODO: implement constant folding for arithmetics, string concat
                // Value::constant_fold(self, other) -> Option<Value>?
                // need to pass the op as well...^
                self.compile_expr(op.left);
                self.compile_expr(op.right);
                let instr = match op.op {
                    BinaryOperator::Mul => Instruction::Mul,
                    BinaryOperator::Div => Instruction::Div,
                    BinaryOperator::Mod => Instruction::Mod,
                    BinaryOperator::Plus => Instruction::Plus,
                    BinaryOperator::Minus => Instruction::Minus,
                    BinaryOperator::FloorDiv => Instruction::FloorDiv,
                    BinaryOperator::Power => Instruction::Power,
                    BinaryOperator::LessThan => Instruction::LessThan,
                    BinaryOperator::GreaterThan => Instruction::GreaterThan,
                    BinaryOperator::LessThanOrEqual => Instruction::LessThanOrEqual,
                    BinaryOperator::GreaterThanOrEqual => Instruction::GreaterThanOrEqual,
                    BinaryOperator::Equal => Instruction::Equal,
                    BinaryOperator::NotEqual => Instruction::NotEqual,
                    BinaryOperator::StrConcat => Instruction::StrConcat,
                    BinaryOperator::In => Instruction::In,
                    BinaryOperator::And => todo!("what do we generate"),
                    BinaryOperator::Or => todo!("what do we generate"),
                    // These are not really binops and we already switched them to separate AST
                    // nodes in the parser
                    BinaryOperator::Is | BinaryOperator::Pipe => unreachable!(),
                };
                self.chunk.add_instruction_with_span(instr, span);
            }
        }
    }

    pub fn compile_node(&mut self, node: Node) {
        match node {
            Node::Content(text) => {
                self.chunk.add_instruction(Instruction::WriteText(text));
            }
            Node::Expression(expr) => {
                self.compile_expr(expr);
                self.chunk.add_instruction(Instruction::WriteTop);
            }
            Node::Set(s) => {
                self.compile_expr(s.value);
                let instr = if s.global {
                    Instruction::SetGlobal(s.name)
                } else {
                    Instruction::Set(s.name)
                };
                self.chunk.add_instruction(instr);
            }
            Node::Include(i) => {
                self.chunk.add_instruction(Instruction::Include(i.name));
            }
            Node::Block(_) => {}
            Node::ForLoop(_) => {}
            Node::If(_) => {}
            Node::FilterSection(_) => {}
        }
    }

    pub(crate) fn test_compile_tmp(&mut self, nodes: Vec<Node>) {
        println!("{:?}", nodes);
        for node in nodes {
            self.compile_node(node);
        }
    }
}
