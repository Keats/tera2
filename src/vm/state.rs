use crate::parsing::Chunk;
use crate::vm::for_loop::ForLoop;
use crate::vm::stack::Stack;
use crate::{Context, Value};

use crate::utils::Span;
use std::collections::BTreeMap;

/// The state of the interpreter.
/// We pass it around rather than put it on the VM to avoid multiple borrow issues
/// when dealing with inheritance.
#[derive(Debug)]
pub(crate) struct State<'tera> {
    pub(crate) stack: Stack<'tera>,
    pub(crate) chunk: &'tera Chunk,
    pub(crate) for_loops: Vec<ForLoop>,
    /// Any variables with {% set %} outside a for loop or {% set_global %} will be stored here
    /// Locals set in a for loop are set in `for_loops`
    set_variables: BTreeMap<String, Value>,
    pub(crate) context: &'tera Context,
    /// To handle the capture instructions
    pub(crate) capture_buffers: Vec<Vec<u8>>,
    /// Used in includes only
    pub(crate) include_parent: Option<&'tera State<'tera>>,

    /// (block name, (all_chunks, level))
    pub(crate) blocks: BTreeMap<&'tera str, (Vec<&'tera Chunk>, usize)>,
    pub(crate) current_block_name: Option<&'tera str>,
}

impl<'t> State<'t> {
    pub(crate) fn new(context: &'t Context, chunk: &'t Chunk) -> Self {
        Self {
            stack: Stack::new(),
            for_loops: Vec::with_capacity(4),
            set_variables: BTreeMap::new(),
            context,
            chunk,
            capture_buffers: Vec::with_capacity(4),
            include_parent: None,
            blocks: BTreeMap::new(),
            current_block_name: None,
        }
    }

    pub(crate) fn current_tpl_name(&self) -> &str {
        &self.chunk.name
    }

    pub(crate) fn store_local(&mut self, name: &str, value: Value) {
        if let Some(forloop) = self.for_loops.last_mut() {
            forloop.store(name, value);
        } else {
            self.store_global(name, value);
        }
    }

    pub(crate) fn store_global(&mut self, name: &str, value: Value) {
        self.set_variables.insert(name.to_string(), value);
    }

    /// Loads the value with the current name on the stack
    /// It goes in the following order for scopes:
    /// 1. All loops from the inner to the outer
    /// 2. set_variables
    /// 3. self.context
    /// 4. return Value::Undefined
    pub(crate) fn get(&self, name: &str) -> Value {
        for forloop in &self.for_loops {
            if let Some(v) = forloop.get(name) {
                return v;
            }
        }

        if let Some(val) = self.set_variables.get(name) {
            return val.clone();
        }

        if let Some(val) = self.context.data.get(name) {
            return val.clone();
        }

        if let Some(parent) = self.include_parent {
            parent.get(name)
        } else {
            Value::Undefined
        }
    }

    pub(crate) fn load_name<'tera: 't>(&mut self, name: &str, span: &'tera Option<Span>) {
        self.stack
            .push_borrowed(self.get(name), span.as_ref().unwrap());
    }
}
