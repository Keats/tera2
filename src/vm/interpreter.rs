use std::collections::BTreeMap;
use std::io::Write;

use crate::errors::{Error, TeraResult};
use crate::parsing::compiler::CompiledMacroDefinition;
use crate::parsing::{Chunk, Instruction};
use crate::template::Template;
use crate::value::Value;
use crate::vm::for_loop::ForLoop;

use crate::vm::state::State;
use crate::{Context, Tera};

// TODO: add a method to error that automatically adds span + template source
pub(crate) struct VirtualMachine<'tera> {
    tera: &'tera Tera,
    template: &'tera Template,
}

impl<'tera> VirtualMachine<'tera> {
    pub fn new(tera: &'tera Tera, template: &'tera Template) -> Self {
        Self { tera, template }
    }

    // TODO: do that at compile time so we don't need to do it at runtime
    fn get_block_lineage(&self, block_name: &str) -> TeraResult<Vec<&'tera Chunk>> {
        // We first get all the chunks we might need to render
        let mut blocks = Vec::with_capacity(10);
        // The block is present in the template we are rendering
        if let Some(bl) = self.template.block_lineage.get(block_name) {
            for c in bl {
                blocks.push(c);
            }
        } else {
            // the block is not present, we go up the lineage by 1 until we find a template that has it
            for parent_tpl_name in self.template.parents.iter().rev() {
                let parent_tpl = self.tera.get_template(parent_tpl_name)?;
                if let Some(bl) = parent_tpl.block_lineage.get(block_name) {
                    for c in bl {
                        blocks.push(c);
                    }
                    break;
                }
            }
        }

        Ok(blocks)
    }

    fn interpret(&self, state: &mut State<'tera>, output: &mut impl Write) -> TeraResult<()> {
        let mut ip = 0;
        macro_rules! op_binop {
            ($op:tt) => {{
                let b = state.stack.pop();
                let a = state.stack.pop();
                state.stack.push(Value::from(a $op b));
            }};
        }

        macro_rules! math_binop {
            ($fn:ident) => {{
                let b = state.stack.pop();
                let a = state.stack.pop();
                state.stack.push(crate::value::number::$fn(&a, &b)?);
            }};
        }

        macro_rules! write_to_buffer {
            ($val:expr) => {{
                if let Some(captured) = state.capture_buffers.last_mut() {
                    write!(captured, "{}", $val)?;
                } else {
                    write!(output, "{}", $val)?;
                }
            }};
        }

        // TODO later: blocks/tests/filters/fns
        // println!("{:?}", self.template.chunk);
        while let Some(instr) = state.chunk.get(ip) {
            // println!("{}. {:?}", state.chunk.name, instr);
            match instr {
                Instruction::LoadConst(v) => state.stack.push(v.clone()),
                Instruction::LoadName(n) => state.load_name(n),
                Instruction::LoadAttr(attr) => {
                    let a = state.stack.pop();
                    if a == Value::Undefined {
                        return Err(Error::message(
                            "Found undefined. TODO: point to right variable/span by walking the back the stack".into(),
                        ));
                    }
                    state.stack.push(a.get_attr(attr));
                }
                Instruction::BinarySubscript => {
                    let subscript = state.stack.pop();
                    let val = state.stack.pop();
                    if val == Value::Undefined {
                        return Err(Error::message(
                            "Found undefined. TODO: point to right variable/span by walking the back the stack".into(),
                        ));
                    }
                    state.stack.push(val.get_item(subscript));
                }
                Instruction::WriteText(t) => {
                    write_to_buffer!(t);
                }
                Instruction::WriteTop => {
                    let top = state.stack.pop();
                    if top == Value::Undefined {
                        return Err(Error::message(
                            "Found undefined. TODO: point to right variable/span by walking the back the stack".into(),
                        ));
                    }
                    write_to_buffer!(top);
                }
                Instruction::Set(name) => {
                    let val = state.stack.pop();
                    state.store_local(name, val);
                }
                Instruction::SetGlobal(name) => {
                    let val = state.stack.pop();
                    state.store_global(name, val);
                }
                Instruction::Include(name) => {
                    self.render_include(name, state, output)?;
                }
                Instruction::BuildMap(num_elem) => {
                    let mut elems = Vec::with_capacity(*num_elem);
                    for _ in 0..*num_elem {
                        let val = state.stack.pop();
                        let key = state.stack.pop();
                        elems.push((key.as_key()?, val));
                    }
                    let map: BTreeMap<_, _> = elems.into_iter().collect();
                    state.stack.push(Value::from(map))
                }
                Instruction::BuildList(num_elem) => {
                    let mut elems = Vec::with_capacity(*num_elem);
                    for _ in 0..*num_elem {
                        elems.push(state.stack.pop());
                    }
                    elems.reverse();
                    state.stack.push(Value::from(elems));
                }
                Instruction::CallFunction(fn_name) => {
                    let kwargs = state.stack.pop();
                    if fn_name == "super" {
                        let current_block_name =
                            state.current_block_name.expect("no current block");
                        let (blocks, level) = state
                            .blocks
                            .remove(current_block_name)
                            .expect("no lineage found");
                        if blocks.len() == 1 {
                            return Err(Error::message(
                                "Tried to use super() in the top level block".into(),
                            ));
                        }

                        let block_chunk = blocks[level + 1];
                        let old_chunk = std::mem::replace(&mut state.chunk, block_chunk);
                        state.blocks.insert(current_block_name, (blocks, level + 1));
                        let res = self.interpret(state, output);
                        state.chunk = old_chunk;
                        res?;
                        state.stack.push(Value::Null);
                    } else {
                        println!("Calling {fn_name} with {kwargs:?}");
                    }
                }
                Instruction::ApplyFilter(_) => {}
                Instruction::RunTest(_) => {}
                Instruction::RenderMacro(idx) => {
                    let kwargs = state.stack.pop().into_map().expect("to have kwargs");
                    let mut context = Context::new();
                    // TODO: inheritance loads macros
                    // see cargo test tests::macros::can_load_macro_in_child
                    // first need to make sure the data in the template makes sense
                    let curr_template = if self.template.parents.is_empty() {
                        self.template
                    } else {
                        self.tera.get_template(state.current_tpl_name())?
                    };
                    // println!("{:?}", curr_template);
                    let compiled_macro_def = &curr_template.macro_calls_def[*idx];
                    for (key, value) in &compiled_macro_def.kwargs {
                        match kwargs.get(&key.as_str().into()) {
                            Some(kwarg_val) => {
                                context.insert(key, kwarg_val);
                            }
                            None => match value {
                                Some(kwarg_val) => {
                                    context.insert(key, kwarg_val);
                                }
                                None => todo!("Missing arg macro error"),
                            },
                        }
                    }

                    let val = self.render_macro(
                        &curr_template.macro_calls[*idx],
                        compiled_macro_def,
                        context,
                    )?;
                    state.stack.push(Value::from(val));
                }
                Instruction::RenderBlock(block_name) => {
                    let block_lineage = self.get_block_lineage(block_name)?;
                    let block_chunk = block_lineage[0];
                    let old_chunk = std::mem::replace(&mut state.chunk, block_chunk);
                    state.blocks.insert(block_name, (block_lineage, 0));
                    let old_block_name =
                        std::mem::replace(&mut state.current_block_name, Some(block_name));
                    let res = self.interpret(state, output);
                    state.chunk = old_chunk;
                    state.current_block_name = old_block_name;
                    res?;
                }
                Instruction::Jump(target_ip) => {
                    ip = *target_ip;
                    continue;
                }
                Instruction::PopJumpIfFalse(target_ip) => {
                    let val = state.stack.pop();
                    if !val.is_truthy() {
                        ip = *target_ip;
                        continue;
                    }
                }
                Instruction::JumpIfFalseOrPop(target_ip) => {
                    let peeked = state.stack.peek();
                    if !peeked.is_truthy() {
                        ip = *target_ip;
                        continue;
                    } else {
                        state.stack.pop();
                    }
                }
                Instruction::JumpIfTrueOrPop(target_ip) => {
                    let peeked = state.stack.peek();
                    if peeked.is_truthy() {
                        ip = *target_ip;
                        continue;
                    } else {
                        state.stack.pop();
                    }
                }
                Instruction::Capture => {
                    state.capture_buffers.push(Vec::with_capacity(128));
                }
                Instruction::EndCapture => {
                    let captured = state.capture_buffers.pop().unwrap();
                    let val = Value::from(String::from_utf8(captured)?);
                    state.stack.push(val);
                }
                Instruction::StartIterate(is_key_value) => {
                    let container = state.stack.pop();
                    state.for_loops.push(ForLoop::new(*is_key_value, container));
                }
                Instruction::Iterate(end_ip) => {
                    // TODO: something very slow happening with forloop
                    if let Some(for_loop) = state.for_loops.last_mut() {
                        if for_loop.is_over() {
                            ip = *end_ip;
                            continue;
                        }
                        for_loop.advance();
                        for_loop.end_ip = *end_ip;
                    } else {
                        unreachable!()
                    }
                }
                Instruction::StoreLocal(name) => {
                    if let Some(for_loop) = state.for_loops.last_mut() {
                        for_loop.store_local(name.as_str());
                    } else {
                        unreachable!()
                    }
                }
                Instruction::PopLoop => {
                    state.for_loops.pop();
                }
                Instruction::Mul => math_binop!(mul),
                Instruction::Div => math_binop!(div),
                Instruction::FloorDiv => math_binop!(floor_div),
                Instruction::Mod => math_binop!(rem),
                Instruction::Plus => math_binop!(add),
                Instruction::Minus => math_binop!(sub),
                Instruction::Power => math_binop!(pow),
                Instruction::LessThan => op_binop!(<),
                Instruction::GreaterThan => op_binop!(>),
                Instruction::LessThanOrEqual => op_binop!(<=),
                Instruction::GreaterThanOrEqual => op_binop!(>=),
                Instruction::Equal => op_binop!(==),
                Instruction::NotEqual => op_binop!(!=),
                Instruction::StrConcat => {
                    let b = state.stack.pop();
                    let a = state.stack.pop();
                    // TODO: we could push_str if `a` is a string
                    state.stack.push(Value::from(format!("{a}{b}")));
                }
                Instruction::In => {
                    let b = state.stack.pop();
                    let a = state.stack.pop();
                    state.stack.push(Value::Bool(b.contains(&a)?));
                }
                Instruction::Not => {
                    let a = state.stack.pop();
                    state.stack.push(Value::from(!a.is_truthy()));
                }
                Instruction::Negative => {
                    let a = state.stack.pop();
                    state.stack.push(crate::value::number::negate(&a)?);
                }
            }

            ip += 1;
        }

        Ok(())
    }

    fn render_macro(
        &self,
        info: &(String, String),
        macro_def: &CompiledMacroDefinition,
        context: Context,
    ) -> TeraResult<String> {
        let tpl = self.tera.get_template(&info.0)?;
        let vm = Self {
            tera: self.tera,
            template: tpl,
        };

        let mut state = State::new(&context, &macro_def.chunk);
        let mut output = Vec::with_capacity(1024);
        vm.interpret(&mut state, &mut output)?;

        Ok(String::from_utf8(output)?)
    }

    fn render_include(
        &self,
        name: &str,
        state: &State<'tera>,
        output: &mut impl Write,
    ) -> TeraResult<()> {
        let tpl = self.tera.get_template(name)?;
        let vm = Self {
            tera: self.tera,
            template: tpl,
        };

        // We create a dummy state for variables to be written to, but we don't keep it around
        let mut include_state = State::new(state.context, &tpl.chunk);
        include_state.include_parent = Some(state);
        vm.interpret(&mut include_state, output)?;
        Ok(())
    }

    pub(crate) fn render(&mut self, context: &Context) -> TeraResult<String> {
        let mut output = Vec::with_capacity(self.template.size_hint());
        // TODO(perf): can we optimize this at the bytecode level to avoid hashmap lookups?
        let chunk = if let Some(base_tpl_name) = self.template.parents.first() {
            let tpl = self.tera.get_template(base_tpl_name)?;
            &tpl.chunk
        } else {
            &self.template.chunk
        };
        let mut state = State::new(context, chunk);

        self.interpret(&mut state, &mut output)?;
        Ok(String::from_utf8(output)?)
    }
}
