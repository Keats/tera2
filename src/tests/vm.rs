use crate::parsing::parser::Parser;
use crate::vm::Compiler;

#[test]
fn vm_ok() {
    insta::glob!("vm_inputs/success/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let nodes = Parser::new(&contents).parse().unwrap();
        let mut compiler = Compiler::new(&path.file_name().unwrap().to_string_lossy(), "");
        compiler.compile(nodes);

        insta::assert_debug_snapshot!(compiler.chunk);
    });
}

#[test]
fn vm_blocks() {
    insta::glob!("vm_inputs/blocks/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        let nodes = Parser::new(&contents).parse().unwrap();
        let mut compiler = Compiler::new(&path.file_name().unwrap().to_string_lossy(), "");
        compiler.compile(nodes);

        let mut s = String::with_capacity(1000);
        s.push_str(&format!("{:?}", compiler.chunk));
        s.push_str("\n\n");

        let mut blocks: Vec<_> = compiler.blocks.into_iter().collect();
        blocks.sort_by(|a, b| a.0.cmp(&b.0));
        for (name, chunk) in blocks {
            s.push_str(&format!(">> Block: {name}\n"));
            s.push_str(&format!("{chunk:?}"));
            s.push_str("---\n");
        }
        insta::assert_snapshot!(s);
    });
}
