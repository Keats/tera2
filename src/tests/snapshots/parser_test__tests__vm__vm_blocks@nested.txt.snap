---
source: src/tests/vm.rs
expression: s
input_file: src/tests/vm_inputs/blocks/nested.txt
---
=== nested.txt ===
0000 CallBlock("hello")


>> Block: hello
=== nested.txt ===
0000 WriteText("body")
0001 CallBlock("inside_hello")
---
>> Block: inside_hello
=== nested.txt ===
0000 WriteText("more")
---

